File: java/src/main/java/codingdojo/Fizzbuzz5.java
package codingdojo;

import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Fizzbuzz5 {

    public static void fizzbuzz() {
        calculateFizzbuzz().forEach(System.out::println);
    }

    static Stream<String> calculateFizzbuzz() {
        IntStream range = IntStream.range(1, 101);
        return range.mapToObj(Fizzbuzz5::fizzbuzzFactor);
    }

    static String fizzbuzzFactor(int n) {
        StringBuilder result = new StringBuilder();
        for (Factor factor : factors) {
            if (factor.predicate.test(n)) {
                result.append(factor.name);
            }
        }
        String fizzbuzz = result.toString();
        return "".equals(fizzbuzz) ? "" + n : fizzbuzz;
    }

    private static List<Factor> factors = Arrays.asList(
            new Factor("Fizz", 3),
            new Factor("Buzz", 5)
    );

    static class Factor {
        public String name;
        public Predicate<Integer> predicate;
        public Factor(String name, int factor) {
            this.name = name;
            this.predicate = n -> n % factor == 0;
        }

    }

}

File: kotlin/src/main/kotlin/codingdojo/Fizzbuzz5.kt
package codingdojo;

import java.util.Arrays;
import java.util.function.Predicate;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class Fizzbuzz5 {
    companion object {
        fun fizzbuzz() {
            calculateFizzbuzz().forEach(System.out::println);
        }

        fun calculateFizzbuzz(): List<String> {
            return (1..100).map { n -> fizzbuzzFactor(n) }
        }

        fun fizzbuzzFactor(n: Int): String {
            val result = StringBuilder();
            for (factor in factors) {
                if (factor.predicate(n)) {
                    result.append(factor.name);
                }
            }
            val fizzbuzz = result.toString();
            if ("".equals(fizzbuzz))
                return "$n"
            else
                return fizzbuzz
        }

        val factors: List<Factor> = listOf(Factor("Fizz", 3), Factor("Buzz", 5));
    }

    class Factor(val name: String, val factor: Int) {
        val predicate = { n: Int -> n % factor == 0 }
    }

}

